Unit test failing

[aries-framework/did-exchange] 2019/10/28 11:36:06 UTC - didexchange.(*Operation).writeResponse -> ERROR Unable to send error response, failed to write
panic: close of closed channel

goroutine 6493 [running]:
github.com/hyperledger/aries-framework-go/pkg/restapi/operation/didexchange.TestServiceEvents.func1(0xb15a09, 0xb, 0xc0020a3560, 0x120, 0x120, 0x43d4c0, 0xc0026aae00)
	/home/soluchok/projects/aries-framework-go/pkg/restapi/operation/didexchange/didexchange_test.go:336 +0x20a
github.com/hyperledger/aries-framework-go/pkg/restapi/operation/didexchange.(*mockNotifier).Notify(0xc0000b9a80, 0xb15a09, 0xb, 0xc0020a3560, 0x120, 0x120, 0x0, 0x431deb)
	/home/soluchok/projects/aries-framework-go/pkg/restapi/operation/didexchange/didexchange_test.go:471 +0x80
github.com/hyperledger/aries-framework-go/pkg/restapi/operation/didexchange.(*Operation).sendConnectionNotification(0xc002346cd0, 0xc002684d50, 0x24, 0x7f7ca045a008, 0xc002341dc0)
	/home/soluchok/projects/aries-framework-go/pkg/restapi/operation/didexchange/didexchange.go:416 +0x5e4
github.com/hyperledger/aries-framework-go/pkg/restapi/operation/didexchange.(*Operation).handleMessageEvents(0xc002346cd0, 0xc002703f18, 0x2, 0x1)
	/home/soluchok/projects/aries-framework-go/pkg/restapi/operation/didexchange/didexchange.go:378 +0x182
github.com/hyperledger/aries-framework-go/pkg/restapi/operation/didexchange.(*Operation).startClientEventListener.func1(0xc002346cd0)
	/home/soluchok/projects/aries-framework-go/pkg/restapi/operation/didexchange/didexchange.go:362 +0x1fe
created by github.com/hyperledger/aries-framework-go/pkg/restapi/operation/didexchange.(*Operation).startClientEventListener
	/home/soluchok/projects/aries-framework-go/pkg/restapi/operation/didexchange/didexchange.go:356 +0x291
FAIL	github.com/hyperledger/aries-framework-go/pkg/restapi/operation/didexchange	1.907s

--- a/pkg/client/didexchange/client.go
+++ b/pkg/client/didexchange/client.go
@@ -174,6 +174,20 @@ func (c *Client) GetConnection(connectionID string) (*Connection, error) {
 	}, nil
 }
 
+// GetConnectionAtState fetches connection record for connection id at particular state.
+func (c *Client) GetConnectionAtState(connectionID, stateID string) (*Connection, error) {
+	conn, err := c.connectionStore.GetConnectionRecordAtState(connectionID, stateID)
+	if err != nil {
+		if errors.Is(err, storage.ErrDataNotFound) {
+			return nil, ErrConnectionNotFound
+		}
+		return nil, fmt.Errorf("cannot fetch state from store: connectionid=%s err=%s", connectionID, err)
+	}
+	return &Connection{
+		conn,
+	}, nil
+}
+
 // RemoveConnection removes connection record for given id
 func (c *Client) RemoveConnection(id string) error {
 	// TODO https://github.com/hyperledger/aries-framework-go/issues/553 RemoveConnection from did exchange service

--- a/pkg/client/didexchange/client_test.go
+++ b/pkg/client/didexchange/client_test.go
@@ -6,7 +6,6 @@ SPDX-License-Identifier: Apache-2.0
 package didexchange
 
 import (
-	"crypto"
 	"encoding/json"
 	"errors"
 	"fmt"
@@ -207,6 +206,23 @@ func TestClient_GetConnection(t *testing.T) {
 	})
 }
 
+func TestClientGetConnectionAtState(t *testing.T) {
+	// create service
+	svc, err := didexchange.New(&mockcreator.MockDIDCreator{}, &mockprotocol.MockProvider{})
+	require.NoError(t, err)
+	require.NotNil(t, svc)
+
+	// create client
+	c, err := New(&mockprovider.Provider{StorageProviderValue: mockstore.NewMockStoreProvider(),
+		ServiceValue: svc})
+	require.NoError(t, err)
+
+	// not found
+	result, err := c.GetConnectionAtState("id1", "complete")
+	require.Equal(t, err.Error(), ErrConnectionNotFound.Error())
+	require.Nil(t, result)
+}
+
 func TestClient_RemoveConnection(t *testing.T) {
 	svc, err := didexchange.New(&mockcreator.MockDIDCreator{}, &mockprotocol.MockProvider{})
 	require.NoError(t, err)
@@ -266,12 +282,11 @@ func TestClient_QueryConnectionsByParams(t *testing.T) {
 
 func TestServiceEvents(t *testing.T) {
 	store := mockstore.NewMockStoreProvider()
-	recorder := didexchange.NewConnectionRecorder(store.Store)
 	didExSvc, err := didexchange.New(&mockcreator.MockDIDCreator{}, &mockprotocol.MockProvider{StoreProvider: store})
 	require.NoError(t, err)
 
 	// create the client
-	c, err := New(&mockprovider.Provider{StorageProviderValue: mockstore.NewMockStoreProvider(), ServiceValue: didExSvc})
+	c, err := New(&mockprovider.Provider{StorageProviderValue: store, ServiceValue: didExSvc})
 	require.NoError(t, err)
 	require.NotNil(t, c)
 
@@ -287,9 +302,13 @@ func TestServiceEvents(t *testing.T) {
 	mCh := make(chan service.StateMsg, 10)
 	err = c.RegisterMsgEvent(mCh)
 	require.NoError(t, err)
+
+	stateMsg := make(chan service.StateMsg)
 	go func() {
 		for e := range mCh {
-			fmt.Println("message = ", e.Type)
+			if e.Type == service.PostState && e.StateID == "responded" {
+				stateMsg <- e
+			}
 		}
 	}()
 
@@ -316,35 +335,22 @@ func TestServiceEvents(t *testing.T) {
 	err = didExSvc.HandleInbound(msg)
 	require.NoError(t, err)
 
-	validateState(t, recorder, id, "responded", 100*time.Millisecond)
-}
-
-func validateState(t *testing.T, recorder *didexchange.ConnectionRecorder, id, expected string,
-	timeoutDuration time.Duration) {
-	actualState := ""
-	theirPrefix := "their"
-	timeout := time.After(timeoutDuration)
-	for {
-		select {
-		case <-timeout:
-			require.Fail(t, fmt.Sprintf("id=%s expectedState=%s actualState=%s", id, expected, actualState))
-			return
+	select {
+	case e := <-stateMsg:
+		switch v := e.Properties.(type) {
+		case Event:
+			props := v
+			conn, err := c.GetConnectionAtState(props.ConnectionID(), e.StateID)
+			require.NoError(t, err)
+			require.Equal(t, e.StateID, conn.State)
 		default:
-			key := createNSKey(theirPrefix, id)
-			connRec, err := recorder.GetConnectionRecordByNSThreadID(key)
-			if err != nil || expected != connRec.State {
-				continue
-			}
-			return
+			require.Fail(t, "unable to cast to did exchange event")
 		}
+	case <-time.After(5 * time.Second):
+		require.Fail(t, "tests are not validated due to timeout")
 	}
 }
-func createNSKey(prefix, id string) string {
-	h := crypto.SHA256.New()
-	hash := h.Sum([]byte(id))
-	storeKey := fmt.Sprintf("%x", hash)
-	return fmt.Sprintf("%s_%s", prefix, storeKey)
-}
+
 func TestServiceEventError(t *testing.T) {
 	didExSvc := mockprotocol.MockDIDExchangeSvc{
 		ProtocolName:           didexchange.DIDExchange,

--- a/pkg/didcomm/protocol/didexchange/persistence.go
+++ b/pkg/didcomm/protocol/didexchange/persistence.go
@@ -114,7 +114,19 @@ func (c *ConnectionRecorder) GetInvitation(id string) (*Invitation, error) {
 
 // GetConnectionRecord return connection record based on the connection ID
 func (c *ConnectionRecorder) GetConnectionRecord(connectionID string) (*ConnectionRecord, error) {
-	k := connectionKeyPrefix(connectionID)
+	return c.getAndUnmarshal(connectionKeyPrefix(connectionID))
+}
+
+// GetConnectionRecordAtState return connection record based on the connection ID and state.
+func (c *ConnectionRecorder) GetConnectionRecordAtState(connectionID, stateID string) (*ConnectionRecord, error) {
+	if stateID == "" {
+		return nil, errors.New("stateID can't be empty")
+	}
+
+	return c.getAndUnmarshal(connectionKeyPrefix(connectionID + stateID))
+}
+
+func (c *ConnectionRecorder) getAndUnmarshal(k string) (*ConnectionRecord, error) {
 	connRecordBytes, err := c.store.Get(k)
 	if err != nil {
 		return nil, fmt.Errorf("get connection record: %w", err)
@@ -139,12 +151,26 @@ func (c *ConnectionRecorder) GetConnectionRecordByNSThreadID(nsThreadID string) 
 }
 
 // saveConnectionRecord saves the connection record against the connection id  in the store
+// TODO - https://github.com/hyperledger/aries-framework-go/issues/622 Transient v Permanent Data store
 func (c *ConnectionRecorder) saveConnectionRecord(record *ConnectionRecord) error {
-	k := connectionKeyPrefix(record.ConnectionID)
-	bytes, err := json.Marshal(record)
+	if err := c.marshalAndSave(connectionKeyPrefix(record.ConnectionID), record); err != nil {
+		return fmt.Errorf("save connection record: %w", err)
+	}
+
+	if record.State != "" {
+		return c.marshalAndSave(connectionKeyPrefix(record.ConnectionID+record.State), record)
+	}
+
+	return nil
+}
+
+func (c *ConnectionRecorder) marshalAndSave(k string, v *ConnectionRecord) error {
+	bytes, err := json.Marshal(v)
+
 	if err != nil {
 		return fmt.Errorf("save connection record: %w", err)
 	}
+
 	return c.store.Put(k, bytes)
 }
 
@@ -179,6 +205,7 @@ func (c *ConnectionRecorder) saveNSThreadID(thid, namespace, connectionID string
 
 	return c.store.Put(k, []byte(connectionID))
 }
+
 func prepareConnectionRecord(connRecBytes []byte) (*ConnectionRecord, error) {
 	connRecord := &ConnectionRecord{}
 	err := json.Unmarshal(connRecBytes, connRecord)

--- a/pkg/didcomm/protocol/didexchange/persistence_test.go
+++ b/pkg/didcomm/protocol/didexchange/persistence_test.go
@@ -158,6 +158,41 @@ func TestConnectionRecorder_GetConnectionRecord(t *testing.T) {
 	})
 }
 
+func TestConnectionRecordByState(t *testing.T) {
+	store := mockstorage.NewMockStoreProvider()
+	record := NewConnectionRecorder(store.Store)
+	require.NotNil(t, record)
+
+	connRec := &ConnectionRecord{ConnectionID: generateRandomID(), ThreadID: threadIDValue,
+		Namespace: myNSPrefix, State: "requested"}
+	err := record.saveConnectionRecord(connRec)
+	require.NoError(t, err)
+
+	// data exists
+	storedConnRec, err := record.GetConnectionRecordAtState(connRec.ConnectionID, "requested")
+	require.NoError(t, err)
+	require.Equal(t, storedConnRec, connRec)
+
+	// data doesn't exists
+	_, err = record.GetConnectionRecordAtState(connRec.ConnectionID, "invalid")
+	require.Error(t, err)
+	require.Contains(t, err.Error(), "get connection record: data not found")
+
+	// data with no state details
+	connRec = &ConnectionRecord{ConnectionID: generateRandomID(), ThreadID: threadIDValue,
+		Namespace: myNSPrefix}
+	err = record.saveConnectionRecord(connRec)
+	require.NoError(t, err)
+	_, err = record.GetConnectionRecordAtState(connRec.ConnectionID, "requested")
+	require.Error(t, err)
+	require.Contains(t, err.Error(), "get connection record: data not found")
+
+	// get with empty stateID
+	_, err = record.GetConnectionRecordAtState(connRec.ConnectionID, "")
+	require.Error(t, err)
+	require.Contains(t, err.Error(), "stateID can't be empty")
+}
+
 func TestConnectionRecorder_SaveConnectionRecord(t *testing.T) {
 	t.Run("save connection record and get connection Record success", func(t *testing.T) {
 		store := &mockstorage.MockStore{Store: make(map[string][]byte)}
@@ -201,6 +236,7 @@ func TestConnectionRecorder_SaveConnectionRecord(t *testing.T) {
 		require.Contains(t, err.Error(), "get error")
 	})
 }
+
 func TestConnectionRecorder_GetConnectionRecordByNSThreadID(t *testing.T) {
 	t.Run(" get connection record by namespace threadID in my namespace", func(t *testing.T) {
 		store := &mockstorage.MockStore{Store: make(map[string][]byte)}
@@ -243,6 +279,7 @@ func TestConnectionRecorder_GetConnectionRecordByNSThreadID(t *testing.T) {
 		require.Nil(t, connRec)
 	})
 }
+
 func TestConnectionRecorder_PrepareConnectionRecord(t *testing.T) {
 	t.Run(" prepare connection record  error", func(t *testing.T) {
 		store := &mockstorage.MockStore{Store: make(map[string][]byte)}
@@ -253,6 +290,7 @@ func TestConnectionRecorder_PrepareConnectionRecord(t *testing.T) {
 		require.Nil(t, connRec)
 	})
 }
+
 func TestConnectionRecorder_CreateNSKeys(t *testing.T) {
 	t.Run(" creating their namespace key success", func(t *testing.T) {
 		key, err := createNSKey(theirNSPrefix, threadIDValue)
@@ -264,6 +302,7 @@ func TestConnectionRecorder_CreateNSKeys(t *testing.T) {
 		require.Contains(t, err.Error(), "empty bytes")
 	})
 }
+
 func TestConnectionRecorder_SaveNSThreadID(t *testing.T) {
 	t.Run("missing required parameters", func(t *testing.T) {
 		store := &mockstorage.MockStore{Store: make(map[string][]byte)}

--- a/pkg/restapi/operation/didexchange/didexchange.go
+++ b/pkg/restapi/operation/didexchange/didexchange.go
@@ -375,7 +375,7 @@ func (c *Operation) handleMessageEvents(e service.StateMsg) error {
 		switch v := e.Properties.(type) {
 		case didexchange.Event:
 			props := v
-			err := c.sendConnectionNotification(props.ConnectionID())
+			err := c.sendConnectionNotification(props.ConnectionID(), e.StateID)
 			if err != nil {
 				return fmt.Errorf("send connection notification failed : %w", err)
 			}
@@ -393,8 +393,8 @@ func (c *Operation) handleActionEvents(e service.DIDCommAction) {
 	e.Continue()
 }
 
-func (c *Operation) sendConnectionNotification(connectionID string) error {
-	conn, err := c.client.GetConnection(connectionID)
+func (c *Operation) sendConnectionNotification(connectionID, stateID string) error {
+	conn, err := c.client.GetConnectionAtState(connectionID, stateID)
 	if err != nil {
 		logger.Errorf("Send notification failed, topic[%s], connectionID[%s]", connectionsWebhookTopic, connectionID)
 		return fmt.Errorf("connection notification webhook : %w", err)

--- a/pkg/restapi/operation/didexchange/didexchange_test.go
+++ b/pkg/restapi/operation/didexchange/didexchange_test.go
@@ -15,7 +15,6 @@ import (
 	"io"
 	"net/http"
 	"net/http/httptest"
-	"sync"
 	"testing"
 	"time"
 
@@ -322,9 +321,6 @@ func TestServiceEvents(t *testing.T) {
 	require.NoError(t, err)
 
 	done := make(chan struct{})
-	// TODO: the test is failing sometimes, need to review the test logic
-	//       and make sure that `once` solution is fine or fix this test according to logic
-	var once sync.Once
 
 	// create the client
 	op, err := New(&mockprovider.Provider{StorageProviderValue: store,
@@ -338,7 +334,7 @@ func TestServiceEvents(t *testing.T) {
 				require.NoError(t, jsonErr)
 
 				if conn.State == "responded" {
-					once.Do(func() { close(done) })
+					close(done)
 				}
 
 				return nil
@@ -419,13 +415,10 @@ func TestHandleMessageEvent(t *testing.T) {
 	require.Error(t, err)
 	require.Contains(t, err.Error(), "service processing failed : err type")
 
-	op.notifier = &mockNotifier{notifyFunc: func(topic string, message []byte) error {
-		return errors.New("webhook error")
-	}}
 	err = op.handleMessageEvents(service.StateMsg{Type: service.PostState, Properties: &didExEvent{}})
 	require.Error(t, err)
 	require.Contains(t, err.Error(), "send connection notification failed : "+
-		"connection notification webhook : webhook error")
+		"connection notification webhook :")
 }
 
 func TestSendConnectionNotification(t *testing.T) {
@@ -438,20 +431,21 @@ func TestSendConnectionNotification(t *testing.T) {
 	connBytes, err := json.Marshal(connRec)
 	require.NoError(t, err)
 	require.NoError(t, storeProv.Store.Put("conn_id1", connBytes))
+	require.NoError(t, storeProv.Store.Put("conn_id1"+"completed", connBytes))
 	t.Run("send notification success", func(t *testing.T) {
 		op, err := New(&mockprovider.Provider{StorageProviderValue: storeProv,
 			ServiceValue: &protocol.MockDIDExchangeSvc{}}, webhook.NewHTTPNotifier(nil))
 		require.NoError(t, err)
-		err = op.sendConnectionNotification(connID)
+		err = op.sendConnectionNotification(connID, "completed")
 		require.NoError(t, err)
 	})
 	t.Run("send notification connection not found error", func(t *testing.T) {
 		op, err := New(&mockprovider.Provider{StorageProviderValue: storeProv,
 			ServiceValue: &protocol.MockDIDExchangeSvc{}}, webhook.NewHTTPNotifier(nil))
 		require.NoError(t, err)
-		err = op.sendConnectionNotification("id2")
+		err = op.sendConnectionNotification("id2", "")
 		require.Error(t, err)
-		require.Contains(t, err.Error(), "connection notification webhook : connection not found")
+		require.Contains(t, err.Error(), "connection notification webhook : cannot fetch state from store:")
 	})
 	t.Run("send notification webhook error", func(t *testing.T) {
 		op, err := New(&mockprovider.Provider{StorageProviderValue: storeProv,
@@ -462,7 +456,7 @@ func TestSendConnectionNotification(t *testing.T) {
 		})
 		require.NoError(t, err)
 		require.NotNil(t, op)
-		err = op.sendConnectionNotification(connID)
+		err = op.sendConnectionNotification(connID, "completed")
 		require.Error(t, err)
 		require.Contains(t, err.Error(), "connection notification webhook : webhook error")
 	})

